components/odh-notebook-controller/controllers/notebook_webhook.go

```diff
@@ -268,137 +263,86 @@ func (w *NotebookWebhook) Handle(ctx context.Context, req admission.Request) adm
 	return nil
 }
 
-func InjectProxyConfig(notebook *nbv1.Notebook) error {
-	notebookContainers := &notebook.Spec.Template.Spec.Containers
-	var imgContainer corev1.Container
+func CheckAndMountCACertBundle(ctx context.Context, cli client.Client, notebook *nbv1.Notebook) error {
 
-	// Add trusted-ca volume
-	notebookVolumes := &notebook.Spec.Template.Spec.Volumes
-	certVolumeExists := false
-	certVolume := corev1.Volume{
-		Name: "trusted-ca",
-		VolumeSource: corev1.VolumeSource{
-			ConfigMap: &corev1.ConfigMapVolumeSource{
-				LocalObjectReference: corev1.LocalObjectReference{
-					Name: "trusted-ca",
-				},
-				Items: []corev1.KeyToPath{{
-					Key:  "ca-bundle.crt",
-					Path: "tls-ca-bundle.pem",
-				},
-				},
-			},
-		},
-	}
-	for index, volume := range *notebookVolumes {
-		if volume.Name == "trusted-ca" {
-			(*notebookVolumes)[index] = certVolume
-			certVolumeExists = true
-			break
-		}
-	}
-	if !certVolumeExists {
-		*notebookVolumes = append(*notebookVolumes, certVolume)
+	configMapName := "odh-trusted-ca-bundle"
+
+	// Fetch the list of ConfigMaps from the cluster
+	configMapList := &corev1.ConfigMapList{}
+	if err := cli.List(ctx, configMapList); err != nil {
+		return err
 	}
 
-	// Update Notebook Image container with env variables and Volume Mounts
-	for _, container := range *notebookContainers {
-		// Update notebook image container with env Variables
-		if container.Name == notebook.Name {
-			var newVars []corev1.EnvVar
-			imgContainer = container
-
-			for key, val := range proxyEnvVars {
-				keyExists := false
-				for _, env := range imgContainer.Env {
-					if key == env.Name {
-						keyExists = true
-						// Update if Proxy spec is updated
-						if env.Value != val {
-							env.Value = val
-						}
-					}
-				}
-				if !keyExists {
-					newVars = append(newVars, corev1.EnvVar{Name: key, Value: val})
-				}
+	// Search for the odh-trusted-ca-bundle ConfigMap
+	for i := range configMapList.Items {
+		cm := &configMapList.Items[i]
+		if cm.Name == configMapName {
+
+			// Check if the odh-ca-bundle.crt key exists
+			certData, ok := cm.Data["odh-ca-bundle.crt"]
+			if !ok {
+				return fmt.Errorf("key 'odh-ca-bundle.crt' not found in ConfigMap '%s'", configMapName)
 			}
 
-			// Update container only when required env variables are not present
-			imgContainerExists := false
-			if len(newVars) != 0 {
-				imgContainer.Env = append(imgContainer.Env, newVars...)
+			// Attempt to decode PEM encoded certificate
+			block, _ := pem.Decode([]byte(certData))
+			if block != nil && block.Type == "CERTIFICATE" {
+				// Attempt to parse the certificate
+				_, err := x509.ParseCertificate(block.Bytes)
+				if err != nil {
+					return fmt.Errorf("error parsing certificate for key 'odh-ca-bundle.crt' in ConfigMap '%s': %v", configMapName, err)
+				}
+			} else if len(certData) > 0 {
+				return fmt.Errorf("invalid certificate format for key 'odh-ca-bundle.crt' in ConfigMap '%s'", configMapName)
 			}
```

Comment:
This checks only the first certificate in bundle. If the bundle contains more than one certificate after the first one, and some of them are malformed, this check will not discover that.

Reply:
Added the same check for `ca-bundle.crt`, but since it has the ability to auto-heal, I couldn't test it with wrong certificate.

Reply:
That's not what I have in mind, look at this code, see the misformated certificate bundle, your check passes on this bundle

```go
package main

import (
	"crypto/x509"
	"encoding/pem"
	"fmt"
)

func main() {
	certData := `-----BEGIN CERTIFICATE-----
MIIDsDCCAxmgAwIBAgIBATANBgkqhkiG9w0BAQUFADCBnTELMAkGA1UEBhMCVVMx
FzAVBgNVBAgTDk5vcnRoIENhcm9saW5hMRAwDgYDVQQHEwdSYWxlaWdoMRYwFAYD
VQQKEw1SZWQgSGF0LCBJbmMuMQswCQYDVQQLEwJJUzEWMBQGA1UEAxMNUmVkIEhh
dCBJUyBDQTEmMCQGCSqGSIb3DQEJARYXc3lzYWRtaW4tcmR1QHJlZGhhdC5jb20w
HhcNMDkwOTE2MTg0NTI1WhcNMTkwOTE0MTg0NTI1WjCBnTELMAkGA1UEBhMCVVMx
FzAVBgNVBAgTDk5vcnRoIENhcm9saW5hMRAwDgYDVQQHEwdSYWxlaWdoMRYwFAYD
VQQKEw1SZWQgSGF0LCBJbmMuMQswCQYDVQQLEwJJUzEWMBQGA1UEAxMNUmVkIEhh
dCBJUyBDQTEmMCQGCSqGSIb3DQEJARYXc3lzYWRtaW4tcmR1QHJlZGhhdC5jb20w
gZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAN/HDWGiL8BarUWDIjNC6uxCXqYN
QkwcmhILX+cl+YuDDArFL1pYVrith228gF3dSUU5X7kIOmPkkjNheRkbnas61X+n
i3+KWvbX3q+h5VMxKX2cA1U+R3jLuXqYjF+N2gkPyPvxeoDuEncKAItw+mK/r+4L
WBb5nFzek7hP3017AgMBAAGjgf0wgfowHQYDVR0OBBYEFA2sGXDtBKdeeKv+i6g0
6yEmwVY1MIHKBgNVHSMEgcIwgb+AFA2sGXDtBKdeeKv+i6g06yEmwVY1oYGjpIGg
MIGdMQswCQYDVQQGEwJVUzEXMBUGA1UECBMOTm9ydGggQ2Fyb2xpbmExEDAOBgNV
BAcTB1JhbGVpZ2gxFjAUBgNVBAoTDVJlZCBIYXQsIEluYy4xCzAJBgNVBAsTAklT
MRYwFAYDVQQDEw1SZWQgSGF0IElTIENBMSYwJAYJKoZIhvcNAQkBFhdzeXNhZG1p
bi1yZHVAcmVkaGF0LmNvbYIBATAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBBQUA
A4GBAFBgO5y3JcPXH/goumNBW7rr8m9EFZmQyK5gT1Ljv5qaCSZwxkAomhriv04p
mb1y8yjrK5OY3WwgaRaAWRHp4/hn2HWaRvx3S+gwLM7p8V1pWnbSFJOXF3kbuC41
voMIMqAFfHKidKN/yrjJg/1ahIjSt11lMUvRJ4TNT+pk5VnB
-----END CERTIFICATE-----

-----broken nonsense not valid CERTIFICATE-----
MIIDsDCCAxmgAwIBAgIBATANBgkqhkiG9w0BAQUFADCBnTELMAkGA1UEBhMCVVMx
-----END CERTIFICATE-----`

	block, _ := pem.Decode([]byte(certData))
	if block != nil && block.Type == "CERTIFICATE" {
		cert, err := x509.ParseCertificate(block.Bytes)
		if err != nil {
			fmt.Println("error parsing certificate for key 'odh-ca-bundle.crt' in ConfigMap '%s': %v", "xxx", err)
		}
		fmt.Printf("Certificate details: %v\n", cert.Subject)
	} else if len(certData) > 0 {
		fmt.Println("invalid certificate format for key 'odh-ca-bundle.crt' in ConfigMap '%s'", "xxx")
	}

	fmt.Println("Successs!")
}
```

Reply:
Why to print the cert details if success? I don't get the point. We just want to validate if the cert is valid
```
		if err != nil {
			fmt.Println("error parsing certificate for key 'odh-ca-bundle.crt' in ConfigMap '%s': %v", "xxx", err)
		} else {
			fmt.Printf("Certificate details: %v\n", cert.Subject)
			fmt.Println("Successs!")
		}
```

Reply:
@atheo89 that's not the part I wanted to turn your attention to, I wanted to highlight

```
block, _ := pem.Decode([]byte(certData))
````

which loads (and validates) only the first certificate in the file/configmap key, but there can be many certificates after the first one in it, and they are not validated by the current code.

Anyways, the validation runs only the first time the pod is created, but as @harshad explained, the user is able to update the certs afterwards and the change gets propagated. Furthermore, I am guessing the OpenShift Cluster Network operator (https://docs.openshift.com/container-platform/4.14/networking/configuring-a-custom-pki.html#certificate-injection-using-operators_configuring-a-custom-pki) is likely validating the bundle as well.

Let's land this PR first, because it's almost the deadline, and maybe bring this up later, if there is a reason and opportunity?

Reply:
Ahh i see you meant that there is a list of certs. Yeah we definitely check this further and also maybe along with this https://github.com/opendatahub-io/kubeflow/pull/252#discussion_r1508644784